src/Workspaces/Core/Portable/CodeActions/Operations/RenameDocumentOperation.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CodeActions
+{
+    internal class RenameDocumentOperation : CodeActionOperation
+    {
+        private readonly DocumentId _oldDocumentId;
+        private readonly DocumentId _newDocumentId;
+        private readonly string _newFileName;
+        private readonly SourceText _text;
+
+        public override string Title { get; }
+
+        internal override bool ApplyDuringTests => true;
+
+        public RenameDocumentOperation(
+            DocumentId oldDocumentId, 
+            DocumentId newDocumentId, 
+            string newFileName,
+            SourceText text)
+        {
+            _oldDocumentId = oldDocumentId;
+            _newDocumentId = newDocumentId;
+            _newFileName = newFileName;
+            _text = text;
+        }
+
+        internal override bool TryApply(Workspace workspace, IProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            var solution = workspace.CurrentSolution;
+
+            // currently, document rename is accomplished by a remove followed by an add.
```

Comment:
Just had a random other thought that we've probably had a longstanding bug here: I'm guessing for Team Foundation Version Control we won't correctly apply a rename here so we screw up source control history? Or does something else in the workspace figure this out later?